tools/module_selector.py

```diff
@@ -0,0 +1,132 @@
+
+
+def select_modules(registry, selections, random_percentage=None):
+    """
+    Select module versions matching the given patterns and optionally apply a random sample percentage.
+    """
+    selected_modules = []
+
+    for selection in selections:
+        if "@" not in selection:
+            raise ValueError(f"Invalid selection pattern (missing '@'): {selection}")
+
+        module_pattern, version = selection.split("@", 1)
+
+        if not module_pattern or not version:
+            raise ValueError(f"Invalid selection pattern: {selection}")
+
+        regex_pattern = "^" + module_pattern.replace(".", "\\.").replace("*", ".*") + "$"
+        module_regex = re.compile(regex_pattern)
+
+        matching_modules = [module for module in registry.get_all_modules() if module_regex.match(module)]
+
+        for module in matching_modules:
+            module_versions = [m[1] for m in registry.get_module_versions(module)]  # This should be sorted already
+            if version == "latest":
+                latest_version = module_versions[-1]
+                selected_modules.append(f"{module}@{latest_version}")
+            elif version.startswith(">="):
+                selected_modules.extend(
+                    [f"{module}@{v}" for v in module_versions if Version(v) >= Version(version[2:])]
+                )
+            elif version.startswith("<="):
+                selected_modules.extend(
+                    [f"{module}@{v}" for v in module_versions if Version(v) <= Version(version[2:])]
+                )
+            elif version.startswith(">"):
+                selected_modules.extend([f"{module}@{v}" for v in module_versions if Version(v) > Version(version[1:])])
+            elif version.startswith("<"):
+                selected_modules.extend([f"{module}@{v}" for v in module_versions if Version(v) < Version(version[1:])])
+            else:
+                if version in module_versions:
+                    selected_modules.append(f"{module}@{version}")
+                else:
+                    raise ValueError(f"Version {version} of module {module} not found.")
+
+    if random_percentage is not None:
+        percentage = int(random_percentage)
+        if not (0 < percentage <= 100):
+            raise ValueError("Random percentage must be an integer between 1 and 100.")
+
+        total_modules = len(selected_modules)
+        num_to_select = max(1, (percentage * total_modules) // 100)
+        selected_modules = random.sample(selected_modules, num_to_select)
+
+    if not selected_modules:
+        raise ValueError("No matching modules found.")
+
+    return sorted(set(selected_modules))
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Select module versions matching given patterns.")
+    parser.add_argument(
+        "--registry",
+        type=str,
+        default=".",
+        help="Specify the root path of the registry (default: the current working directory).",
```

Comment:
maybe update this to say the workspace root somehow?

Reply:
I mean it could be also a random directory respecting the registry layout? So it doesn't have to be a Bazel workspace root?

Reply:
Ah, I see, you meant when running with Bazel? Updated!